[JSC] Add `JSObject::convertToIndexingTypeIfNeeded`

sosukesuzuki opened the pull request:
#### 3e9dded80a2961f7f37376ea077e34ce7b679908
<pre>
[JSC] Add `JSObject::convertToIndexingTypeIfNeeded`
<a href="https://bugs.webkit.org/show_bug.cgi?id=283816">https://bugs.webkit.org/show_bug.cgi?id=283816</a>

Reviewed by Yusuke Suzuki.

This patch adds `JSObject::convertToIndexingTypeIfNeeded`.
This patch changes to use it for `Array#fill`.

* Source/JavaScriptCore/runtime/JSArray.cpp:
(JSC::JSArray::fastFill):
* Source/JavaScriptCore/runtime/JSObject.cpp:
(JSC::JSObject::convertToIndexingTypeIfNeeded):
* Source/JavaScriptCore/runtime/JSObject.h:

Canonical link: <a href="https://commits.webkit.org/288324@main">https://commits.webkit.org/288324@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/04b1f39434cf01f34e593be6513429c980bd16e0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82505 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/2162 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/36604 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/87631 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33566 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/2233 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/10061 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64296 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/22061 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85567 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1596 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/75058 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44573 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1494 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29246 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/32600 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/75493 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/72786 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29884 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/88993 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/81557 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9804 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/7067 "Found 2 new test failures: http/wpt/mediarecorder/MediaRecorder-video-h264-profiles.html http/wpt/mediarecorder/record-96KHz-sources.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72703 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/10031 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70872 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71919 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17908 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/16058 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/15070 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/1187 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9757 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/15278 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/103965 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/9631 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25219 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/13097 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11401 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/JSArray.cpp:
```diff
@@ -486,44 +486,13 @@ bool JSArray::fastFill(VM& vm, unsigned startIndex, unsigned endIndex, JSValue v
+    if (!(type & IsArray) || hasAnyArrayStorage(type))
+        return false;
+    IndexingType nextType = leastUpperBoundOfIndexingTypeAndValue(type, value);
+    if (type != ArrayWithUndecided && type != nextType)
```
Do we need to limit type == nextType?

sosukesuzuki replied:
Thank you. We don't need it so I removed this branch

Constellation commented on Source/JavaScriptCore/runtime/JSArray.cpp:
```diff
@@ -486,44 +486,11 @@ bool JSArray::fastFill(VM& vm, unsigned startIndex, unsigned endIndex, JSValue v
-            return false;
-        }
-    } else if (type != nextType)
+    if (!(type & IsArray) || hasAnyArrayStorage(type))
```
Probably, We should check hasArrayStorage(nextType) too

sosukesuzuki replied:
Fixed

sosukesuzuki committed:
[JSC] Add `JSObject::convertToIndexingTypeIfNeeded`
https://bugs.webkit.org/show_bug.cgi?id=283816

Reviewed by Yusuke Suzuki.

This patch adds `JSObject::convertToIndexingTypeIfNeeded`.
This patch changes to use it for `Array#fill`.

* Source/JavaScriptCore/runtime/JSArray.cpp:
(JSC::JSArray::fastFill):
* Source/JavaScriptCore/runtime/JSObject.cpp:
(JSC::JSObject::convertToIndexingTypeIfNeeded):
* Source/JavaScriptCore/runtime/JSObject.h:

Canonical link: https://commits.webkit.org/288324@main
